### ISSUE TITLE
Dynamic class name

### ISSUE THREAD

**Issue by Tobbe** [markdown-it/markdown-it-container#23]
Is it possible to have a generic container that applies a dynamic class name?

Kind of like the info-string for fenced code blocks. I.e. if you write `\`\`\`java` you get a `<code>` with the class name `language-java`, but if you write `\`\`\`ruby` the class name will be `language-ruby`.

In a similar manner I would like to be able to do `::: foo` and get `<div class="foo">` and `::: bar` and get `<div class="bar">`. Is this possible? It's important that all use the same opening/closing sequence (`:::`) and that I don't have to write anything more than just those three colons and the classname.

**Comment by Tobbe** [comment]
I think I solved it myself. I wrote a custom render function, like this:

```javascript
md.use(markdownItContainer, 'dynamic', {
    validate: function() { return true; },
    render: function(tokens, idx) {
        var token = tokens[idx];

        if (token.nesting === 1) {
            return '<div class="' + token.info.trim() + '">';
        } else {
            return '</div>';
        }
    },
});
```

Do you see any problems with an implementation like that?

Usage:

```md
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vivamus metus tortor, dignissim a libero sed, rhoncus feugiat neque.

::: foo
In fermentum felis non dapibus hendrerit. Proin et urna viverra, tempor purus a, accumsan nulla.
:::

::: bar baz
Vestibulum ornare eros nisl, in facilisis ante tempor non. Vestibulum sit amet arcu ultricies, sollicitudin urna ut, facilisis enim.
:::

Mauris non efficitur justo, quis consectetur risus. Duis eget pellentesque enim. Integer sagittis, turpis vel semper tincidunt, nunc ante dapibus nibh, sit amet ullamcorper ex metus vitae quam. Aliquam id libero in nulla commodo porttitor sed varius diam. 
```

Output:

```html
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vivamus metus tortor, dignissim a libero sed, rhoncus feugiat neque.</p>

<div class="foo">
<p>In fermentum felis non dapibus hendrerit. Proin et urna viverra, tempor purus a, accumsan nulla.</p>
</div>

<div class="bar baz">
<p>Vestibulum ornare eros nisl, in facilisis ante tempor non. Vestibulum sit amet arcu ultricies, sollicitudin urna ut, facilisis enim.</p>
</div>

<p>Mauris non efficitur justo, quis consectetur risus. Duis eget pellentesque enim. Integer sagittis, turpis vel semper tincidunt, nunc ante dapibus nibh, sit amet ullamcorper ex metus vitae quam. Aliquam id libero in nulla commodo porttitor sed varius diam. </p>
```

**Comment by mb21** [comment]
Actually, you need the validate function. Otherwise it will work only with `::: dynamic`.

I think the behaviour described in this issue would be a great default. It's also how [pandoc behaves by default](http://pandoc.org/MANUAL.html#extension-fenced_divs) now. So this:

```
::: warning
*here be dragons*
:::
```

would work with just:

    var md = require('markdown-it')().use( require('markdown-it-container') );

instead of the current:

    var md = require('markdown-it')().use( require('markdown-it-container'), 'warning' );

**Comment by puzrin** [comment]
I don't agree. You mix params and container (generated tag) type. Those are ortogonal (independent) Type is mandatory.

**Comment by mb21** [comment]
Not sure I understood that. I'm suggesting that if no `container name` is provided to the plugin, this:

```
::: x
content
:::
```

is converted to:

```
<div class="x">
content
</div>
```

for any string `x`.

Are you saying, that you don't agree that this is a sensible default?

But you agree that it _is_ pandoc's default behaviour?

**Comment by mb21** [comment]
Maybe I'm misunderstand the intended behaviour, because at first I tried:

    .use( require('markdown-it-container'), 'div' )

But that didn't work with:

```
::: warning
*here be dragons*
:::
```

**Comment by puzrin** [comment]
My comment was about

```
var md = require('markdown-it')().use( require('markdown-it-container') );
```

instead of

```
var md = require('markdown-it')().use( require('markdown-it-container'), 'warning' );
```

You can set last param any name you like, but it should not be missed.

**Comment by mb21** [comment]
But what if I want my users to be able to use any class name? Which is in fact what pandoc users can do.

**Comment by mb21** [comment]
If you're saying that's a bad default, maybe this could be done by explicitly setting an option, like:

    .use( require('markdown-it-container'), 'div', {
      anyClass: true
    })

**Comment by puzrin** [comment]
I think, you misunderstand the intent of this module. Intent is to be universal (modifyable), for any use case.

If you wish to have default class - just modify renderer to add it, when infostring is empty.

**Comment by mb21** [comment]
I'm not sure whether we're talking past each other here. I don't want to have a default class. I want what's pandoc's behaviour, which is the same what the OP of this issue achieved.

If the only way to do this is with a custom `render` function, that's fine. I just thought it would be nice to do this without that somewhat ugly workaround.

**Comment by cekvenich** [comment]
Tobbe : thanks, I'm using your code at metabake.org for nested css blocks

**Comment by flaviotordini** [comment]
Thanks @Tobbe for the code! It should have been the default behaviour.
